File: Leetcode_Problems/0102.py
# LeetCode 0102 - Binary Tree Level Order Traversal
# Medium - Trees
# Given the root of a binary tree, return the level order traversal of its nodes' values. (i.e., from left to right, level by level).

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# Iterative Breadth-First Search Solution
# Complexities:
# Time : O(n)
# Space: O(n)
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        result = []

        queue = collections.deque()
        queue.append(root)

        while queue:
            queue_length = len(queue)
            level = []
            for i in range(queue_length):
                node = queue.popleft()
                if node:
                    level.append(node.val)
                    queue.append(node.left)
                    queue.append(node.right)
            if level:
                result.append(level)
        return result

File: Leetcode_Problems/0102.go
// LeetCode 0102 - Binary Tree Level Order Traversal
// Medium - Trees
// Given the root of a binary tree, return the level order traversal of its nodes' values. (i.e., from left to right, level by level).

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

// Breadth-First Search Solution
// Complexities:
// Time : O(n)
// Space: O(n)
func levelOrder(root *TreeNode) [][]int {
    result := [][]int{}
    if root == nil { return result }

    queue := []*TreeNode{root}

    for len(queue) > 0 {
        queue_length := len(queue)
        level := []int{}
        
        for i := 0; i < queue_length; i++ {
            node := queue[0]
            queue = queue[1:]
            level = append(level, node.Val)
            if node.Left != nil { queue = append(queue, node.Left)}
            if node.Right != nil { queue = append(queue, node.Right)}
        }
        result = append(result, level)
    }
    return result
}
